In [1]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [2]:
# Load the given datasets
df1 = pd.read_csv('chronic_kidney_disease_categorical.csv')
df2 = pd.read_csv('chronic_kidney_disease_numerical.csv')

# Print the data
df1
df2

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

### Answer:   

The datasets provided for Chronic Kidney Disease (CKD) analysis are divided into two main categories: categorical and numerical data, each representing different types of features that help predict CKD. Below is a breakdown of each:

#### 1. Categorical Dataset:
This dataset contains categorical features, which represent qualitative characteristics or categories. The features either take on binary values (yes or no) or nominal values (e.g, normal or abnormal).

##### Features Overview:
- `unique_id`: A unique identifier for each patient.
- `al`: Albumin levels (recorded as categorical values in this dataset).
- `su`: Sugar levels.
- `rbc`: Red blood cells status (categorical: normal or abnormal).
- `pc`: Pus cells (categorical: normal or abnormal).
- `pcc`: Pus cell clumps (presence or absence).
- `ba`: Bacteria (present or absent).
- `htn`: Hypertension (yes or no).
- `dm`: Diabetes mellitus (yes or no).
- `cad`: Coronary artery disease (yes or no).
- `appet`: Appetite status (good or poor).
- `pe`: Pedal edema (yes or no).
- `ane`: Anemia (yes or no).
- `Target`: The target label, representing whether the person has Chronic Kidney Disease (ckd) or not (notckd).

##### Example of Categorical Columns:
- `rbc`: Red blood cells (normal/abnormal).
- `htn`: Hypertension (yes/no).
- `Target`: Classification of kidney disease (ckd / notckd ).

#### 2. Numerical Dataset:
This dataset contains numerical features, which represent either continuous or discrete quantities, taken during medical examinations and give numerical values that help in assessing a patient's health.

##### Features:
- `unique_id`: A unique identifier for each entry.
- `age`: The age of the patient (years).
- `bp`: Blood pressure (mmHg).
- `bgr`: Blood glucose random (mg/dL).
- `bu`: Blood urea (mg/dL).
- `sc`: Serum creatinine (mg/dL).
- `sod`: Sodium level in the blood (mEq/L).
- `pot`: Potassium level in the blood (mEq/L).
- `hemo`: Hemoglobin level (g/dL).
- `pcv`: Packed cell volume.
- `wbcc`: White blood cell count (cells/cumm).
- `rbcc`: Red blood cell count (millions/cumm).

##### Example of Numerical Columns:
- `age`: Age of the patient (years).
- `bp`: Blood pressure (mmHg).
- `bgr`: Blood glucose random (mg/dL).

#### Types of Features:
- The **categorical dataset** consists of features that represent categories or binary values.
- The **numerical dataset** consists of features which are continuous numerical values.


# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [6]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numerical_duplicates = df2[df2.duplicated()]
print("Numerical dataset - Total duplicate rows:", numerical_duplicates.shape[0])

# Drop these duplicate rows
df2_cleaned = df2.drop_duplicates()


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_duplicates = df1[df1.duplicated()]
print("Categorical dataset - Total duplicate rows:", categorical_duplicates.shape[0])

df1_cleaned = df1.drop_duplicates()



Numerical dataset - Total duplicate rows: 30
Categorical dataset - Total duplicate rows: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

#### Answer:

Not all the unique IDs are present in both datasets. The numerical dataset contains 390 unique IDs, and the categorical dataset contains 400. After merging the numerical and categorical datasets based on the 'unique_id' column, the dataset will contain 390 unique IDs, excluding the 10 outliers, as there is no corresponding numerical data for them.

In [8]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged_data = pd.merge(df2_cleaned, df1_cleaned, on='unique_id', how='inner')

#Print the combined dataset
merged_data


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
388,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [9]:
# Calculate the percentage of rows that contain atleast one missing value
rows_missing_value = merged_data.isnull().any(axis=1).sum()
total_rows = merged_data.shape[0]
percentage_missing = (rows_missing_value / total_rows) * 100

# Print %

print(f"Percentage of rows with at least one missing value: {percentage_missing:.2f}%")

# Drop these rows from the dataset
cleaned_merged_data = merged_data.dropna()

# Print the Dataset
cleaned_merged_data

Percentage of rows with at least one missing value: 60.77%


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
13,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
379,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
381,462892,36.0,80.0,85.0,16.0,1.1,142.0,4.1,15.6,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
386,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,...,abnormal,notpresent,present,yes,yes,no,good,yes,no,ckd
387,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.4: Sort the dataset according to the Labels

In [10]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = cleaned_merged_data.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
sorted_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,...,abnormal,present,present,no,no,no,good,no,yes,ckd
1,137148,73.0,100.0,295.0,90.0,5.6,140.0,2.9,9.2,30.0,...,abnormal,present,notpresent,yes,yes,yes,poor,no,no,ckd
2,484175,53.0,90.0,70.0,107.0,7.2,114.0,3.7,9.5,29.0,...,abnormal,present,notpresent,yes,yes,no,poor,no,yes,ckd
3,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
4,532520,60.0,90.0,105.0,53.0,2.3,136.0,5.2,11.1,33.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61.0,70.0,120.0,29.0,0.7,137.0,3.5,17.4,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,125587,37.0,60.0,109.0,47.0,1.1,141.0,4.9,15.0,48.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,614376,57.0,80.0,119.0,17.0,1.2,135.0,4.7,15.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,853426,41.0,70.0,125.0,38.0,0.6,140.0,5.0,16.8,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [22]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Create dummy variables for categorical columns
encoded_data = pd.get_dummies(sorted_data, columns=categorical_columns)

# Convert boolean true/False to 1/0
encoded_data = encoded_data.astype(int)

# Print the dataset
encoded_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21,90,107,40,1,125,3,8,23,...,1,0,1,0,1,0,0,1,1,0
1,137148,73,100,295,90,5,140,2,9,30,...,0,1,0,1,1,0,1,0,1,0
2,484175,53,90,70,107,7,114,3,9,29,...,1,0,0,1,1,0,0,1,1,0
3,343710,61,80,173,148,3,135,5,7,24,...,0,1,0,1,0,1,0,1,1,0
4,532520,60,90,105,53,2,136,5,11,33,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61,70,120,29,0,137,3,17,52,...,1,0,1,0,1,0,1,0,0,1
149,125587,37,60,109,47,1,141,4,15,48,...,1,0,1,0,1,0,1,0,0,1
150,614376,57,80,119,17,1,135,4,15,42,...,1,0,1,0,1,0,1,0,0,1
151,853426,41,70,125,38,0,140,5,16,41,...,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

##### Answer: 
For binary variables, using a single column (1 for normal, 0 for abnormal), is eqivalent to one-hot encoding, as both convey the same information to our model. 

However, when dealing with categorical variables that have more than two values (e.g., 'type of condiment' with five possible values), using a single column would introduce a false ordinal relationship, which can be potentially confusing for our model. For instance, the model may interpret one value as "greater" than another, which in reality, is false. 

One-hot encoding avoids this issue by treating each category as a separate binary feature, having separate columns for 'ketchup', 'mustard', etc., each indicating the presence or absence of that particular condiment, ensuring the model does not assume any unintended ordering between categories.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [23]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = encoded_data[column].mean()
    std_dev = encoded_data[column].std()
    outlier_threshold_upper = mean + 3 * std_dev
    outlier_threshold_lower = mean - 3 * std_dev
    encoded_data = encoded_data[(encoded_data[column] <= outlier_threshold_upper) & (encoded_data[column] >= outlier_threshold_lower)]

# Print the dataset
encoded_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,21,90,107,40,1,125,3,8,23,...,1,0,1,0,1,0,0,1,1,0
1,137148,73,100,295,90,5,140,2,9,30,...,0,1,0,1,1,0,1,0,1,0
3,343710,61,80,173,148,3,135,5,7,24,...,0,1,0,1,0,1,0,1,1,0
4,532520,60,90,105,53,2,136,5,11,33,...,1,0,1,0,1,0,1,0,1,0
6,546225,46,60,163,92,3,141,4,9,28,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,61,70,120,29,0,137,3,17,52,...,1,0,1,0,1,0,1,0,0,1
149,125587,37,60,109,47,1,141,4,15,48,...,1,0,1,0,1,0,1,0,0,1
150,614376,57,80,119,17,1,135,4,15,42,...,1,0,1,0,1,0,1,0,0,1
151,853426,41,70,125,38,0,140,5,16,41,...,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [24]:
# Normalize the all Numerical Attributes in the dataset.
encoded_data[numerical_columns] = encoded_data[numerical_columns].apply(lambda x: (x - x.min()) / (x.max() - x.min()))

# Print the dataset
encoded_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,474407,0.202703,0.75,0.158798,0.196078,0.125,0.166667,0.333333,0.1,0.000000,...,1,0,1,0,1,0,0,1,1,0
1,137148,0.905405,1.00,0.965665,0.522876,0.625,0.666667,0.000000,0.2,0.225806,...,0,1,0,1,1,0,1,0,1,0
3,343710,0.743243,0.50,0.442060,0.901961,0.375,0.500000,1.000000,0.0,0.032258,...,0,1,0,1,0,1,0,1,1,0
4,532520,0.729730,0.75,0.150215,0.281046,0.250,0.533333,1.000000,0.4,0.322581,...,1,0,1,0,1,0,1,0,1,0
6,546225,0.540541,0.00,0.399142,0.535948,0.375,0.700000,0.666667,0.2,0.161290,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,647552,0.743243,0.25,0.214592,0.124183,0.000,0.566667,0.333333,1.0,0.935484,...,1,0,1,0,1,0,1,0,0,1
149,125587,0.418919,0.00,0.167382,0.241830,0.125,0.700000,0.666667,0.8,0.806452,...,1,0,1,0,1,0,1,0,0,1
150,614376,0.689189,0.50,0.210300,0.045752,0.125,0.500000,0.666667,0.8,0.612903,...,1,0,1,0,1,0,1,0,0,1
151,853426,0.472973,0.25,0.236052,0.183007,0.000,0.666667,1.000000,0.9,0.580645,...,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

#### Answer:

The column 'unique_id' is not appropriate for modeling and predictions, as it is merely an identifier for each row, and does not provide any useful information. 

In [25]:
# Remove that column
encoded_data = encoded_data.drop(columns=['unique_id'])

# Print the dataset
encoded_data

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.125,0.166667,0.333333,0.1,0.000000,0.653226,...,1,0,1,0,1,0,0,1,1,0
1,0.905405,1.00,0.965665,0.522876,0.625,0.666667,0.000000,0.2,0.225806,0.217742,...,0,1,0,1,1,0,1,0,1,0
3,0.743243,0.50,0.442060,0.901961,0.375,0.500000,1.000000,0.0,0.032258,0.395161,...,0,1,0,1,0,1,0,1,1,0
4,0.729730,0.75,0.150215,0.281046,0.250,0.533333,1.000000,0.4,0.322581,0.500000,...,1,0,1,0,1,0,1,0,1,0
6,0.540541,0.00,0.399142,0.535948,0.375,0.700000,0.666667,0.2,0.161290,0.830645,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.743243,0.25,0.214592,0.124183,0.000,0.566667,0.333333,1.0,0.935484,0.217742,...,1,0,1,0,1,0,1,0,0,1
149,0.418919,0.00,0.167382,0.241830,0.125,0.700000,0.666667,0.8,0.806452,0.217742,...,1,0,1,0,1,0,1,0,0,1
150,0.689189,0.50,0.210300,0.045752,0.125,0.500000,0.666667,0.8,0.612903,0.153226,...,1,0,1,0,1,0,1,0,0,1
151,0.472973,0.25,0.236052,0.183007,0.000,0.666667,1.000000,0.9,0.580645,0.161290,...,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [26]:
# Export the dataframe to a new csv file
encoded_data.to_csv('cleaned_chronic_kidney_disease_data.csv', index=False)

# Export the dataframe to a new json file
encoded_data.to_json('cleaned_chronic_kidney_disease_data.json', orient='records', lines=True)


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [27]:
# Output the top 15 rows of the df
encoded_data.head(15)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.202703,0.75,0.158798,0.196078,0.125,0.166667,0.333333,0.1,0.000000,0.653226,...,1,0,1,0,1,0,0,1,1,0
1,0.905405,1.00,0.965665,0.522876,0.625,0.666667,0.000000,0.2,0.225806,0.217742,...,0,1,0,1,1,0,1,0,1,0
3,0.743243,0.50,0.442060,0.901961,0.375,0.500000,1.000000,0.0,0.032258,0.395161,...,0,1,0,1,0,1,0,1,1,0
4,0.729730,0.75,0.150215,0.281046,0.250,0.533333,1.000000,0.4,0.322581,0.500000,...,1,0,1,0,1,0,1,0,1,0
6,0.540541,0.00,0.399142,0.535948,0.375,0.700000,0.666667,0.2,0.161290,0.830645,...,1,0,1,0,1,0,1,0,1,0
8,0.675676,0.75,0.253219,0.633987,0.750,0.366667,0.666667,0.2,0.193548,0.169355,...,1,0,1,0,1,0,1,0,1,0
10,0.716216,0.50,1.000000,0.163399,0.125,0.066667,0.333333,0.3,0.387097,0.532258,...,1,0,0,1,1,0,1,0,1,0
13,0.729730,0.00,0.935622,0.169935,0.125,0.333333,0.333333,0.0,0.064516,0.879032,...,1,0,0,1,1,0,0,1,1,0
17,0.000000,0.00,0.103004,0.372549,0.125,0.500000,0.666667,0.2,0.225806,1.000000,...,1,0,0,1,1,0,1,0,1,0
19,0.878378,0.00,0.206009,0.751634,0.625,0.533333,0.666667,0.4,0.387097,0.879032,...,1,0,0,1,0,1,1,0,1,0


### Paste the markdown table here

|    |       age |    bp |       bgr |         bu |      sc |       sod |       pot |   hemo |       pcv |       wbcc | ... |   cad_no |   cad_yes |   appet_good |   appet_poor |   pe_no |   pe_yes |   ane_no |   ane_yes |   Target_ckd |   Target_notckd |
|----|-----------|-------|-----------|------------|---------|-----------|-----------|--------|-----------|------------|-----|----------|-----------|--------------|---------------|---------|----------|----------|-----------|---------------|-----------------|
| 0  | 0.202703  | 0.75  | 0.158798  | 0.196078   | 0.125   | 0.166667  | 0.333333  | 0.1    | 0.000000  | 0.653226   | ... | 1        | 0         | 1            | 0             | 1       | 0        | 0        | 1         | 1             | 0               |
| 1  | 0.905405  | 1.00  | 0.965665  | 0.522876   | 0.625   | 0.666667  | 0.000000  | 0.2    | 0.225806  | 0.217742   | ... | 0        | 1         | 0            | 1             | 1       | 0        | 1        | 0         | 1             | 0               |
| 3  | 0.743243  | 0.50  | 0.442060  | 0.901961   | 0.375   | 0.500000  | 1.000000  | 0.0    | 0.032258  | 0.395161   | ... | 0        | 1         | 0            | 1             | 0       | 1        | 0        | 1         | 1             | 0               |
| 4  | 0.729730  | 0.75  | 0.150215  | 0.281046   | 0.250   | 0.533333  | 1.000000  | 0.4    | 0.322581  | 0.500000   | ... | 1        | 0         | 1            | 0             | 1       | 0        | 1        | 0         | 1             | 0               |
| 6  | 0.540541  | 0.00  | 0.399142  | 0.535948   | 0.375   | 0.700000  | 0.666667  | 0.2    | 0.161290  | 0.830645   | ... | 1        | 0         | 1            | 0             | 1       | 0        | 1        | 0         | 1             | 0               |
| 8  | 0.675676  | 0.75  | 0.253219  | 0.633987   | 0.750   | 0.366667  | 0.666667  | 0.2    | 0.193548  | 0.169355   | ... | 1        | 0         | 1            | 0             | 1       | 0        | 1        | 0         | 1             | 0               |
| 10 | 0.716216  | 0.50  | 1.000000  | 0.163399   | 0.125   | 0.066667  | 0.333333  | 0.3    | 0.387097  | 0.532258   | ... | 1        | 0         | 0            | 1             | 1       | 0        | 1        | 0         | 1             | 0               |
| 13 | 0.729730  | 0.00  | 0.935622  | 0.169935   | 0.125   | 0.333333  | 0.333333  | 0.0    | 0.064516  | 0.879032   | ... | 1        | 0         | 0            | 1             | 1       | 0        | 0        | 1         | 1             | 0               |
| 17 | 0.000000  | 0.00  | 0.103004  | 0.372549   | 0.125   | 0.500000  | 0.666667  | 0.2    | 0.225806  | 1.000000   | ... | 1        | 0         | 0            | 1             | 1       | 0        | 1        | 0         | 1             | 0               |
| 19 | 0.878378  | 0.00  | 0.206009  | 0.751634   | 0.625   | 0.533333  | 0.666667  | 0.4    | 0.387097  | 0.879032   | ... | 1        | 0         | 0            | 1             | 0       | 1        | 1        | 0         | 1             | 0               |
| 20 | 0.851351  | 0.25  | 0.618026  | 0.562092   | 0.750   | 0.000000  | 0.333333  | 0.2    | 0.161290  | 0.580645   | ... | 0        | 1         | 1            | 0             | 0       | 1        | 0        | 1         | 1             | 0               |
| 21 | 0.878378  | 0.25  | 0.639485  | 0.470588   | 0.375   | 0.433333  | 0.666667  | 0.3    | 0.322581  | 0.104839   | ... | 0        | 1         | 1            | 0             | 1       | 0        | 1        | 0         | 1             | 0               |
| 22 | 0.783784  | 0.00  | 0.725322  | 0.313725   | 0.500   | 0.566667  | 1.000000  | 0.2    | 0.193548  | 0.258065   | ... | 1        | 0         | 0            | 1             | 0       | 1        | 1        | 0         | 1             | 0               |
| 24 | 0.662162  | 0.50  | 0.618026  | 0.411765   | 0.375   | 0.566667  | 0.666667  | 0.3    | 0.354839  | 0.250000   | ... | 1        | 0         | 1            | 0             | 0       | 1        | 1        | 0         | 1             | 0               |
| 25 | 0.770270  | 1.00  | 0.901288  | 0.163399   | 0.375   | 0.766667  | 0.333333  | 0.6    | 0.548387  | 0.443548   | ... | 0        | 1         | 1            | 0             | 1       | 0        | 1        | 0         | 1             | 0               |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

#### Response:

When utilizing LLMs for data conversion, one can mitigate hallucinations through a number of strategies, including:

1. Segmenting Data: Break larger datasets into smaller segments for focused processing, effectively minimizing the cognitive load on the model.
2.  Iterative Refining: Continuously refine and validate the output of the LLM to ensure accuracy
3. Manual Review: Regularly review the output of the LLM to identify and correct any inaccuracies


## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [28]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc': 'Red Blood Cells',
    'pc': 'Pus Cell',
    'pcc': 'Pus Cell clumps',
    'ba': 'Bacteria',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rc': 'Red Blood Cell Count',
    'htn': 'Hypertension',
    'dm': 'Diabetes Mellitus',
    'cad_no': 'Coronary Artery Disease (No)',
    'cad_yes': 'Coronary Artery Disease (Yes)',
    'appet_good': 'Appetite (Good)',
    'appet_poor': 'Appetite (Poor)',
    'pe_no': 'Pedal Edema (No)',
    'pe_yes': 'Pedal Edema (Yes)',
    'ane_no': 'Anemia (No)',
    'ane_yes': 'Anemia (Yes)',
    'Target_ckd': 'Class (CKD)',
    'Target_notckd': 'Class (Not CKD)'
}

encoded_data.rename(columns=rename_mapping, inplace=True)

encoded_data

,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,...,Coronary Artery Disease (No),Coronary Artery Disease (Yes),Appetite (Good),Appetite (Poor),Pedal Edema (No),Pedal Edema (Yes),Anemia (No),Anemia (Yes),Class (CKD),Class (Not CKD)
0,0.202703,0.75,0.158798,0.196078,0.125,0.166667,0.333333,0.1,0.000000,0.653226,...,1,0,1,0,1,0,0,1,1,0
1,0.905405,1.00,0.965665,0.522876,0.625,0.666667,0.000000,0.2,0.225806,0.217742,...,0,1,0,1,1,0,1,0,1,0
3,0.743243,0.50,0.442060,0.901961,0.375,0.500000,1.000000,0.0,0.032258,0.395161,...,0,1,0,1,0,1,0,1,1,0
4,0.729730,0.75,0.150215,0.281046,0.250,0.533333,1.000000,0.4,0.322581,0.500000,...,1,0,1,0,1,0,1,0,1,0
6,0.540541,0.00,0.399142,0.535948,0.375,0.700000,0.666667,0.2,0.161290,0.830645,...,1,0,1,0,1,0,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,0.743243,0.25,0.214592,0.124183,0.000,0.566667,0.333333,1.0,0.935484,0.217742,...,1,0,1,0,1,0,1,0,0,1
149,0.418919,0.00,0.167382,0.241830,0.125,0.700000,0.666667,0.8,0.806452,0.217742,...,1,0,1,0,1,0,1,0,0,1
150,0.689189,0.50,0.210300,0.045752,0.125,0.500000,0.666667,0.8,0.612903,0.153226,...,1,0,1,0,1,0,1,0,0,1
151,0.472973,0.25,0.236052,0.183007,0.000,0.666667,1.000000,0.9,0.580645,0.161290,...,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [29]:
# Group by the 'Class (CKD)' and 'Class (Not CKD)' columns and count occurrences
result_ckd = encoded_data.groupby('Class (CKD)').size().reset_index(name='count')
result_notckd = encoded_data.groupby('Class (Not CKD)').size().reset_index(name='count')

# Concatenate results into one DataFrame
result = pd.concat([result_ckd.rename(columns={'Class (CKD)': 'Target'}),
                    result_notckd.rename(columns={'Class (Not CKD)': 'Target'})], ignore_index=True)

# Display the resulting DataFrame
result

,Target,count
0,0,111
1,1,22
2,0,22
3,1,111


*Describe what the above code does here*

#### Answer:
The provided code counts the occurrences of two classes, 'Class (CKD)' and 'Class (Not CKD)', in a pandas DataFrame. It groups the data by these classes, calculates the respective counts, and resets the index to create a unified format. The results are combined into a single DataFrame with a 'Target' column representing the classes and a 'count' column showing the counts for each class. Finally, it prints this combined DataFrame for easy interpretation of the class distribution.
